### DISCUSSION TITLE
GrandOrgue code and IDEs

### DISCUSSION THREAD

**Discussion by larspalo** [GrandOrgue/grandorgue#771] [Developers]
Currently the GO source code contain a /nbproject folder that's not essential to the GO source. In the .gitignore file only /nbproject/private is listed thus making any changes done in the /nbproject under version control. This is highly undesirable, I think. Personally I mostly use a simple text editor for coding, but perhaps we need to adopt a strategy that's suitable to different developers taste in IDEs and their different project files.

I suggest we remove the /nbproject folder and it's content from GrandOrgue source code. Next we agree upon a way in which developers should have their project files separate from the source code, and possibly add some entries to .gitignore that will cover all needed cases.

**Comment by oleg68** [comment]
Because creating a project for IDE is not quite simple and may cause errors I prefer to include projects for different IDEs to the source tree.

Personally I work with GO on different computers and keeping netbeans project in the repo solves the problem of synchronisation.

**Reply by larspalo** [reply]
But not in separate folders in the root directory I hope. At least we should have a structure like /ide/netbeans, /ide/eclipse etc.

You won't like it as much when I or someone else edit the files to suite our setup... We need to find a better way!

In my opinion, only IDE project files that can be imported and used by anyone without user specific data should reside in the GO source tree.

**Reply by oleg68** [reply]
The netbeans project files are rellocatable: the project may be cloned to any directory.

**Reply by larspalo** [reply]
Good. Still, your personal files should be in your fork and not in the official repository.

**Reply by larspalo** [reply]
Can you make an export of the netbeans project that others can actually import? That might have a place in the official repository.

**Comment by oleg68** [comment]
Personal files are in the netbeans/private subdirectory and they are ignored by git. Everyone can open the progect in his own environment without any problems.

Netbeans does not have any special 'export' feature because the same  project can be opened by other people in every personal environment.

**Reply by larspalo** [reply]
Have you looked into the project.xml and configurations.xml files of /nbproject? They both have:

`/home/oleg/my-projects/grandorgue/GrandOrgue`

and the likes of paths in them. Hardly useful for anyone else, right? I even tested opening them in netbeans and there's no useful way I can see to import or open the project from that folder or the content in it.

So, I don't mind having some "ide_projects" folder in the source tree that's completely ignored in the official .gitignore and each developer can put whatever suits them there. But those files should not make their way into the official repository!

And yes, it seems that it's possible to export a project as a .zip from netbeans, but as the project folder is now it's useless as far as I can see.

**Comment by oleg68** [comment]
> `/home/oleg/my-projects/grandorgue/GrandOrgue`
> 
> and the likes of paths in them. Hardly useful for anyone else, right?

It does not prevent using this project files by anyone in any other directory path.

> but as the project folder is now it's useless as far as I can see.

Netbeans calculates relative pathes from absolute paths of the files and the project root folder. Futher it uses only the relative paths and ignores the absolute paths at all.

I agree that does not look very nice, but it is still working.

I've made a context replacement of `/home/oleg/my-projects/grandorgue/GrandOrgue` to `.` in both project.xml and configurations.xml. The project remained working. May be commiting these files after such replacement would be an acceptable compromise?

> At least we should have a structure like /ide/netbeans, /ide/eclipse etc.

Unfortunally most of IDEs require the predefined directory structure beyong the root project directory and will not work from `ide` subdirectory.

> And yes, it seems that it's possible to export a project as a .zip from netbeans,

1. I couldn't find such option
2. If even the option existed, placing a zip file to the git repository wouldn't be a good idea

**Reply by larspalo** [reply]
@oleg68 So, how would anyone actually use the netbeans project files? I couldn't find any way in the Netbeans 10.0 included in Ubuntu 20.04, but I admit to knowing next to nothing about netbeans - I just looked into the file menu options and tested them to no avail.

I've easily imported my local fork into eclipse as other project using git though - it seems to have some nice git features integrated. Remains to get the build scripts setup working from within eclipse, I've not spent so much time on it yet and I'm not too excited about it either as I'm fairly old-school in my programming approach. And yes, annoyingly, eclipse places a .project file in the root of the repository. This file is easily hidden by adding it to the .git/info/exclude file that is **not** under version control and nothing needs to be added to the .gitignore file in the root.

Unless you show how the netbeans project files can benefit anyone else than you I still think they should be removed from the official repository.

**Reply by oleg68** [reply]
> So, how would anyone actually use the netbeans project files? I

1. Clone GrandOrgue from github to a local directory. For example, `~/GrandOrgue`

    ```
    git clone --recurse-submodules git@github.com:GrandOrgue/grandorgue.git ~/GrandOrgue
    ```
    it will create a directory ~/GrandOrgue. The `nbproject` will be inside.

2. Start netbeans
3. Make sure that C/C++ plugin is installed and is activated (Tools -> Plugins)
4. `File` -> `Open Project...` and select a directory `~/GrandOrgue` or another one where you have clone the repo. But **not `~/GrandOrgue/nbproject`**
5. After the project is open you can do with the right-click context menu on the `GrandOrgue` project in the 'Projects' tab

    1. More Build Commands -> `PreBuild`. It executes cmake
    2. Build. It executes make
    3. Run. Launches GrandOrgue has already build
    4. Debug. Launches GrandOrgue with the internal debugger.

> h. And yes, annoyingly, eclipse places a .project file in the root of the repository. This file is easily hidden by adding it to the .git/info/exclude file that is not under version control and nothing needs to be added to the .gitignore file in the root.

If you use eclipse, yor are welcome to put the eclipse project to the repository.

> Unless you show how the netbeans project files can benefit anyone else than you I still think they should be removed from the official repository.

My approach is to include projects for various IDEs in the source tree. It will make involving new people quite easier.

**Reply by oleg68** [reply]
> I couldn't find any way in the Netbeans 10.0 included in Ubuntu 20.04

I use Netbeans 12.3 and 12.4 on different computers. I'm not sure that the project saved with netbeans 12.x can be easy opened with 10x. But maybe it works.

**Reply by oleg68** [reply]
> Remains to get the build scripts setup working from within eclipse

It is not a very good using the IDE because our build scripts can build GO from scratch and do not support incremental compilation. But incremental build save much time when you modify and debug only a few source files.

So the better approach is to integrate cmake and make to the IDE.

> either as I'm fairly old-school in my programming approach.

If you do not like IDEs at all, you may not use them and may ignore existence of the IDE files in the source tree. But please do not make the life of the developers that use IDEs harder.

Of cause, the IDE files should be workable with any people on every environment.

**Reply by oleg68** [reply]
> Unless you show how the netbeans project files can benefit anyone else than you I still think they should be removed from the official repository.

I don't want to enforce all developers to use only netbeans. Selecting a certain IDE is a thing of a flavor.

But if we are talking about recruiting of new developers that it is malprobably that they are also old-school. Almost all young people prefer using IDEs. My approach is to allow them to put basic settings of their favorite IDEs to the repository.

**Reply by larspalo** [reply]
> So the better approach is to integrate cmake and make to the IDE.

Yes, that should be relatively easy since it's already working as it should. I'll test it later with eclipse.

At least with netbeans 10.0 I couldn't get your instructions to work, I've also tested with 12.5 that's also available for Ubuntu 20.04 but it just complains about unrecognized project missing plugin? And I have c/c++ active. Some other?

**Comment by rousseldenis** [comment]
Personally, I'd prefer to not include specific development environment stuffs in project.

As some maybe want to have their specific configuration.

In all the projects I've involved in, it's more common to have entries in .gitignore like :

.vscode/
.env
.idea/

Constraining external tools in the project is generally a bad idea.

**Reply by oleg68** [reply]
> Personally, I'd prefer to not include specific development environment stuffs in project.

It makes setting up IDEs very hard.

**Reply by rousseldenis** [reply]
But it's up to developer flavor. We cannot embed here every IDE's configuration as there are so much

**Reply by rousseldenis** [reply]
But what we can do is to do something in wiki or ...

**Reply by oleg68** [reply]
> But it's up to developer flavor. We cannot embed here every IDE's configuration as there are so much

We shouldn't embed **every** existing IDEs, but only ones that are used by GrandOrgue developers. The number of developers is not very high; the number of used IDEs is less.

**Comment by oleg68** [comment]
@larspalo I've just tested using my netbeans project in a fresh Ubuntu 18.04 by `administrator` account. It's home dir is `/home/administrator` that differs from mine.

I wasn't able to do it with netbeans 10, but I was able to do it with netbeans 12.5.

Steps to reproduce:

1. Get a fresh ubuntu
2. Install necessary packages:

    ```
     sudo apt-get install openjdk-11-jdk git
    ```

3. Clone the GrandOrgue repository

    ```
    git clone https://github.com/GrandOrgue/grandorgue.git ~/GrandOrgueSrc
    ```

4. Install dependencies necessary for building GrandOrgue

    ```
    ~/GrandOrgueSrc/build-scripts/for-linux/prepare-debian-ubuntu.sh
    ```

5. Install Netbeans

    1. Launch `Application Manager`
    2. Search netbeans
    3. Select `Apache Netbeans`
    4. Click on Install. The version 12.5 is installed

6. Install and activate C/C++ Plugin

    1. Start NetBeans
    2. Tools -> Plugins
    3. Go to `Settings` tab
    4. Enable `NetBeans 8.2 Plugin Portal`
    5. Go to ``Available Plugins`` tab
    6. Click to `Check for Newewst`. `C/C++` will appear in the list of plugins
    7. Select `C/C++`. Click `Install`. Accept the license.
    8. After installing go to the `Installed` tab. Make sure that `C/C++` is activated
    9. Restart Netbeans

7. Open and work with GrandOrgue

    1. File -> Open Project
    2. Select ~/GrandOrgueSrc. The project `GrandOrgue will appear in the `Projects` window.
    3. Right-click on `GrandOrgue`. Select `Other Build Commands` -> Prepare Build. Cmake output will appear
    4. Right-click on `GrandOrgue`. Select `Build`. The compilation progress will appear
    5. Click to the green rectangle on the toolbar. GrandOrgue will be launched
    6. You can also run GO with debugger

**Reply by larspalo** [reply]
@oleg68 Well, I could get it to "kind of" work with the instructions above. I could get netbeans to open the project but it reports warnings and errors - likely due to the absolute paths. So I get warnings for the folders like: `Folder's content is used by project, but physically absent on disk`. And for the .cpp files (I don't see any .h files) I get: `Broken: (absolute path to file) Failed to load project`.

This could likely be fixed by your change in making the paths relative instead.

> The basic idea is to set the project source directory to `../..` in the project.xml.

However, it seems like you have ides/NetBeans12/nbproject/ for the files (three directories down) - why not directly in the ides/NetBeans12 folder (two directories down like above, or did I miss something?

**Comment by oleg68** [comment]
@larspalo 

> At least with netbeans 10.0 I couldn't get your instructions to work, I've also tested with 12.5 that's also available for Ubuntu 20.04 but it just complains about unrecognized project missing plugin? And I have c/c++ active. Some other?

If you have installed C/C++ plugin in nb 10, it does not work in nb 12. You should to uninstall this plugin with nb 10, then run nb 12.5 and install it again. It's better to clear ~/.netbeans directory before start using nb12 otherwise some artifacts from 10 may prevent C/C++ to work with 12.5

**Comment by oleg68** [comment]
@larspalo 

> At least we should have a structure like /ide/netbeans, /ide/eclipse etc.

I managed to move the netbeans project files from the root directory to the `ides/NetBeans12` subrirectory and to get rid of absolute paths in these files. The basic idea is to set the project source directory to `../..` in the project.xml.

Please take a look at https://github.com/oleg68/GrandOrgue-oficial/tree/feature/move-netbeans . If it is OK, I will create a PR.

Maybe eclipse also allows to do such trick?

**Reply by larspalo** [reply]
@oleg68 This sounds reasonable. If the project files actually can be used by any developer and we have a working documentation of it - then ok, they can indeed have a place, I think.

If eclipse also can do such things I cannot tell yet. If it doesn't work in the sub-directory but we still can create a usable project setup, I don't think it's a problem adding it to ides/eclipse together with a readme file (or other documentation) that explain how to move the .project file to root directory and make an entry in the .git/info/exclude file to still be able to use it. I can't say for sure about that yet.

I'll test your instructions above first and see if I can get netbeans to work.

**Comment by oleg68** [comment]
@larspalo 

> Next we agree upon a way in which developers should have their project files separate from the source code, and possibly add some entries to .gitignore that will cover all needed cases.

We are already doing this, arent't we?

> However, it seems like you have ides/NetBeans12/nbproject/ for the files (three directories down) - why not directly in the ides/NetBeans12 folder (two directories down like above, or did I miss something?

Because netbeans requires the subdirectory named `nbproject`, not `NetBeans12` as like as Eclipse requires the the file `.project` in the project directory.

We could create a single project directory for all IDEs `IdeProject` with necessary files and subdirectories for all IDEs `IdeProject/nbproject`, `IdeProject/.project` and so on.

But disadvantages:

1. The mean of the files would not always be evident. For example, it wouldn't be clear that `.project` would be for eclipse.
2. We could not be able to create different project configs for different versions of IDEs, for example, for Netbeans10 and Netbeans12
3. The filename `.project` might cause a conflict with other IDEs.

I thought that the directory name `ides` didn't sound good. May be `ide-projects` is better (`ide-projects/NetBeans12`, `ide-projects/Eclipse/.project`)? What do you thing on the directory name?

>  I don't think it's a problem adding it to ides/eclipse together with a readme file (or other documentation) that explain how to move the .project file to root directory and make an entry in the .git/info/exclude file to still be able to use it.

I prefer to have ready-to use solutions than templates requiring manual steps before using them. You can just open the project `ide-projects/NetBeans12` with netbeans and use it. I'd like to open the project `ide-projects/Eclipse` with eclipse for working with GrandOrgue source code and build it.

> I could get netbeans to open the project but it reports warnings and errors - likely due to the absolute paths. So I get warnings for the folders like: Folder's content is used by project, but physically absent on disk. And for the .cpp files (I don't see any .h files) I get: Broken: (absolute path to file) Failed to load project.

> This could likely be fixed by your change in making the paths relative instead.

Yes, exactly.

**Reply by larspalo** [reply]
> We are already doing this, arent't we?

Yes! That's the very point of this discussion.

> I prefer to have ready-to use solutions than templates requiring manual steps before using them. You can just open the project `ide-projects/NetBeans12` with netbeans and use it. I'd like to open the project `ide-projects/Eclipse` with eclipse for working with GrandOrgue source code and build it.

Indeed, that would be a goal that makes the inclusion of project files in the repository justified as it can perhaps help beginning/new developers on the way. Still there's no replacement for the practical programming experience (coming from doing), but maybe we can make things a little smoother for others to begin the journey.

As for the directory name I don't have any strong feelings really, `ide-projects`, `ide-files` or even just `ide` would be fine for me. I'll go ahead with Eclipse and see what I can be do about it. Also I know that some of the old developers used codeblocks as ide way back in time.

**Comment by oleg68** [comment]
@larspalo So

1. I am remaining `ides` to `ide-projects` and am making a PR for netbeans
2. When you get ready, please make a PR with `ide-projects/Eclipse` or `ide-projects/Eclipse<Version>`

Where might we put the howto-guides for using certain ides (like I posted above about netbeans)? To IDEs.md in the root directory? Or to `ide-projects` subdirectory? Or a separate file for each IDE?

> Also I know that some of the old developers used codeblocks as ide way back in time.

If we do not use it, it is difficult to make a well working project config. So l suggest to let to a new developer that is using codeblocks to do it.

**Comment by oleg68** [comment]
I added #773

**Reply by rousseldenis** [reply]
Great! In a specialized directory is a good idea.
